### PR TITLE
Add `minimize` config for `whitespace` and add handling of `~` in the jinja parser

### LINKED ISSUE #671: Trim all but 1 whitespace jinja condition
@vallentin suggested [here](https://github.com/djc/askama/pull/664#pullrequestreview-945253861) that we could extend the "trim_all_whitespace" feature to be able to trim all whitespace characters but one.

So currently, we have:

```jinja
<div>    {%- if foo -%} blabla {%- endif -%} </div>
```

Which will remove all whitespace characters before and after (by using `-`) and we have:

```jinja
<div>    {%+ if foo +%} blabla {%+ endif +%} </div>
```

Which means that the whitespace should be saved (only useful when the "trim all whitespace" feature is enabled).

The idea would now to add a new character to strip all whitespace characters but one. I propose `~`:

```jinja
<div>    {%~ if foo %} blabla {% endif %} </div>
```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #671.

### CODE REVIEW

File: askama_shared/src/generator.rs [hunk 1]
```diff
@@ -1230,13 +1230,19 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
     fn visit_lit(&mut self, lws: &'a str, val: &'a str, rws: &'a str) {
         assert!(self.next_ws.is_none());
         if !lws.is_empty() {
```
Comment (djc) [inline askama_shared/src/generator.rs:1232]: We should flip the condition and turn it into an early return.

File: askama_shared/src/generator.rs [hunk 2]
```diff
@@ -1230,13 +1230,19 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
...
-            if self.skip_ws {
-                self.skip_ws = false;
+            if self.skip_ws == WhitespaceHandling::Suppress {
```
Comment (djc) [inline askama_shared/src/generator.rs:1233]: Can we do a `match` here somehow? I don't remember exactly how `self.skip_ws` and `val` interact.

File: askama_shared/src/generator.rs [hunk 3]
```diff
@@ -1819,11 +1825,12 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
         self.prepare_ws(ws);
     }
 
-    fn should_trim_ws(&self, ws: Option<Whitespace>) -> bool {
+    fn should_trim_ws(&self, ws: Option<Whitespace>) -> WhitespaceHandling {
         match ws {
-            Some(Whitespace::Trim) => true,
-            Some(Whitespace::Preserve) => false,
-            None => self.whitespace == WhitespaceHandling::Suppress,
+            Some(Whitespace::Trim) => WhitespaceHandling::Suppress,
```
Comment (djc) [inline askama_shared/src/generator.rs:1830]: We should rename `Whitespace::Trim` to `Suppress`, too.

File: askama_shared/src/parser.rs [hunk 4]
```diff
@@ -1447,12 +1447,12 @@ mod tests {
     #[test]
     fn change_delimiters_parse_filter() {
         let syntax = Syntax {
-            expr_start: "{~",
-            expr_end: "~}",
+            expr_start: "{=",
```
[the remark this answers is not in the document]
Comment (djc) [reply askama_shared/src/parser.rs:1450]: A warning is probably a good idea!

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply askama_shared/src/parser.rs:1450]: https://github.com/djc/askama/issues/674 :)

File: askama_shared/src/generator.rs [hunk 5]
```diff
@@ -1230,13 +1230,19 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
...
+            if self.skip_ws == WhitespaceHandling::Suppress {
+                self.skip_ws = WhitespaceHandling::Preserve;
             } else if val.is_empty() {
                 assert!(rws.is_empty());
                 self.next_ws = Some(lws);
-            } else {
+            } else if self.skip_ws == WhitespaceHandling::Preserve {
                 self.buf_writable.push(Writable::Lit(lws));
+            } else if self.skip_ws == WhitespaceHandling::Minimize {
+                if lws.contains('\n') {
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply askama_shared/src/generator.rs:1241]: Sounds good to me!

File: askama_shared/src/generator.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply askama_shared/src/generator.rs:1830]: Adding a new commit then. :)

File: askama_shared/src/generator.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply askama_shared/src/generator.rs:1232]: We can't make it return early, some code after this condition needs to be run.

File: askama_shared/src/generator.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply askama_shared/src/generator.rs]: I replaced this with a `match`. Please tell me if it looks good to you.

**Comment by GuillaumeGomez** [comment]
Updated!

**Review by djc — APPROVED** [review]
LGTM! Any other opinions?

File: askama_shared/src/parser.rs [hunk 6]
```diff
@@ -132,14 +132,16 @@ pub(crate) enum Target<'a> {
 #[derive(Clone, Copy, Debug, PartialEq)]
 pub(crate) enum Whitespace {
```
Comment (vallentin) [inline askama_shared/src/parser.rs:133]: Couldn't `enum WhitespaceHandling` be renamed into `enum Whitespace` and then reuse the same `enum Whitespace` for both things? Is there any real need to have two different types in this case? :)

Comment (GuillaumeGomez) [reply askama_shared/src/parser.rs:133]: I don't like too much to mix parser and config although they are the same type in here. Well, if @djc agrees with you, I'll merge them. :)

Comment (vallentin) [reply askama_shared/src/parser.rs:133]: To be fair, I do agree with that point-of-view. I was just curious if there was anything beyond that :)

Comment (GuillaumeGomez) [reply askama_shared/src/parser.rs:133]: No, just this "mental split". Nothing else. :smile:

Comment (djc) [reply askama_shared/src/parser.rs:133]: The main downside here is the ugly name of `WhitespaceHandling`, but I think it makes sense to keep them separate for now.

**Comment by djc** [comment]
Thanks!

**Pull request merged**